question-servers/calculation-worker.js

```diff
@@ -0,0 +1,202 @@
+        if (err.originalError != null) {
+          e.stack = err.originalError.stack + '\n\n' + err.stack;
+        }
+        reject(e);
+      }
+    );
+  });
+}
+
+function generate(server, coursePath, question, variant_seed) {
+  const questionDir = path.join(coursePath, 'questions', question.directory);
+  const options = question.options || {};
+
+  const questionData = server.getData(variant_seed, options, questionDir);
+  return {
+    params: questionData.params,
+    true_answer: questionData.trueAnswer,
+    options: questionData.options || question.options || {},
+  };
+}
+
+function getFile(server, coursePath, filename, variant, question) {
+  const vid = variant.variant_seed;
+  const params = variant.params;
+  const trueAnswer = variant.true_answer;
+  const options = variant.options;
+  const questionDir = path.join(coursePath, 'questions', question.directory);
+  const fileData = server.getFile(filename, vid, params, trueAnswer, options, questionDir);
+
+  // If `getFile` returns a Buffer, we need to handle that specially, since
+  // Buffers can't be losslessly round-tripped through `JSON.stringify` and
+  // `JSON.parse`.
+  const isBuffer = Buffer.isBuffer(fileData);
+  return {
+    type: isBuffer ? 'buffer' : 'unknown',
+    data: isBuffer ? fileData.toString('base64') : fileData,
+  };
+}
+
+function grade(server, coursePath, submission, variant, question) {
+  const vid = variant.variant_seed;
+  const params = variant.params;
+  const trueAnswer = variant.true_answer;
+  const submittedAnswer = submission.submitted_answer;
+  const options = variant.options;
+  const questionDir = path.join(coursePath, 'questions', question.directory);
+
+  const grading = server.gradeAnswer(
+    vid,
+    params,
+    trueAnswer,
+    submittedAnswer,
+    options,
+    questionDir
+  );
+
+  let score = grading.score;
+  if (!question.partial_credit) {
+    // legacy Calculation questions round the score to 0 or 1 (with 0.5 rounding up)
+    score = grading.score >= 0.5 ? 1 : 0;
+  }
+
+  return {
+    score: score,
+    v2_score: grading.score,
+    feedback: grading.feedback,
+    partial_scores: {},
+    submitted_answer: submission.submitted_answer,
+    format_errors: {},
+    gradable: true,
+    params: variant.params,
+    true_answer: variant.true_answer,
+  };
+}
+
+/**
+ *
+ * @param {import('readline').Interface} rl
+ * @returns {Promise<string | null>}
+ */
+function getLineOnce(rl) {
+  return new Promise((resolve) => {
+    let didResolve = false;
+    rl.on('line', (line) => {
+      if (didResolve) return;
+      didResolve = true;
+      resolve(line);
+    });
+    rl.on('close', () => {
+      if (didResolve) return;
+      didResolve = true;
+      resolve(null);
+    });
+  });
+}
+
+if (require.main === module) {
+  // Redirect `stdout` to `stderr` so that we can ensure that no
+  // user code can write to `stdout`; we need to use `stdout` to send results
+  // instead.
```

Comment:
I think that lots of legacy v2 questions write junk to stderr/stdout, which will now trigger an issue on the question, I think? I'm 100% happy that we will get this junk out of our syslog, but I'm a bit nervous about "breaking" v2 questions in courses that have been "working" for years. I'm almost tempted to silently drop both stderr and stdout here in the interest of just keeping things working?

Reply:
That's fair, yeah. This would mean that there's no easy way for instructors to debug their questions, though I suppose that's ok? The v2 question experience doesn't seem to be worth optimizing at this point.

Reply:
CS 421 has a fair amount of v2 content with sporadic bugs that surfaced in just the past year. We already ran in to dead ends trying to sort that out, so I was thinking it made the most sense for them to migrate content to v3.

Reply:
Upon further reviewing the code, the "error if there's output" piece is actually part of the v3 question processing code. We actually already just ignore the output property on the code caller result (see `question-servers/calculation-subprocess.js#callFunction`), so there's no code change to make here. I'll add a comment explaining that though.

This still means that output will no longer be user-facing; I still think that's ok though.

Reply:
I documented this in 00aaba2.